lib/shelley/src/Cardano/Wallet/Shelley/Network.hs

```diff
@@ -0,0 +1,457 @@
+    )
+import Ouroboros.Network.Protocol.ChainSync.Client
+    ( ChainSyncClient (..)
+    , ClientStIdle (..)
+    , ClientStIntersect (..)
+    , ClientStNext (..)
+    , chainSyncClientPeer
+    )
+import Ouroboros.Network.Protocol.ChainSync.Codec
+    ( codecChainSync )
+import Ouroboros.Network.Protocol.ChainSync.Type
+    ( ChainSync )
+import Ouroboros.Network.Protocol.Handshake.Version
+    ( DictVersion (..), simpleSingletonVersions )
+import Ouroboros.Network.Protocol.LocalTxSubmission.Client
+    ( LocalTxSubmissionClient (..), localTxSubmissionClientPeer )
+import Ouroboros.Network.Protocol.LocalTxSubmission.Codec
+    ( codecLocalTxSubmission )
+import Ouroboros.Network.Protocol.LocalTxSubmission.Type
+    ( LocalTxSubmission )
+
+import Control.Exception
+    ( catch, throwIO )
+import Control.Monad.Class.MonadST
+    ( MonadST )
+import Control.Monad.Class.MonadThrow
+    ( MonadThrow )
+import Control.Monad.Class.MonadTimer
+    ( MonadTimer )
+import Control.Tracer
+    ( Tracer, contramap )
+import Data.ByteString.Lazy
+    ( ByteString )
+import Data.Text
+    ( Text )
+import Data.Void
+    ( Void )
+import Network.Socket
+    ( AddrInfo (..), Family (..), SockAddr (..), SocketType (..) )
+
+import qualified Cardano.Crypto as CC
+import qualified Codec.Serialise as CBOR
+import qualified Network.Socket as Socket
+import qualified Ouroboros.Network.Block as O
+import qualified Ouroboros.Network.Point as Point
+
+-- ----------------
+-- Development
+
+import Cardano.BM.Trace
+    ( nullTracer )
+import Control.Monad
+    ( (>=>) )
+
+-- A point where there is high activity on the mainnet
+point1 :: Point ByronBlock
+point1 =
+    let
+        Right h = CC.decodeAbstractHash "8a7505ec268a934c9c532fcf21185baf849c0de04a6d12cdf80b0fc45fe6ecb3"
+        slot = 21600*19+6540
+
+    in O.Point $ Point.block
+        (O.SlotNo slot)
+        (ByronHash h)
+
+-- How to run the node:
+--
+-- $ git clone https://github.com/input-output-hk/cardano-node
+-- $ cd cardano-node
+-- $ ./scripts/mainnet.sh
+--
+-- Then run main:
+runExperiment
+    :: Point ByronBlock
+       -- ^ Point to start fetching blocks from
+    -> Int
+       -- ^ How many blocks to fetch
+    -> (ByronBlock -> IO ())
+       -- ^ Roll-forward
+    -> (Point ByronBlock -> IO ())
+       -- ^ Roll-backward
+    -> IO ()
+runExperiment start n rollForward rollBackward = do
+    -- I have cardano-node running from ../cardano-node
+    let nodeId = CoreNodeId 0
+    let path = "../cardano-node/socket/" <> (localSocketFilePath nodeId)
+    let addr =  localSocketAddrInfo path
+
+    let params = ChainParameters
+         { epochSlots = EpochSlots 21600
+         , protocolMagic = ProtocolMagicId 764824073
+         }
+
+    let t = nullTracer
+    let client = OuroborosInitiatorApplication $ \pid -> \case
+            ChainSyncWithBlocksPtcl ->
+                chainSyncWithBlocks
+                    pid t params
+                    rollForward
```

Comment:
ok, `rollForward` is not decorated by `Point` as `chainSyncWithBlocks takes it..